ext/soap/php_wss.c

```diff
@@ -0,0 +1,279 @@
+#include "php_soap.h"
+#include "ext/date/php_date.h"
+#include "ext/hash/php_hash.h"
+#include "ext/standard/base64.h"
+#include "ext/standard/php_mt_rand.h"
+#include <libxml/c14n.h>
+#include <libxml/xpathInternals.h>
+
+static zend_string *get_exc_c14n_from_node_ptr(xmlNodePtr node) /* {{{ */
+{
+	zend_string *result = NULL;
+	
+	xmlOutputBufferPtr buffer = xmlAllocOutputBuffer(NULL);
+	if (buffer != NULL) {
+		xmlXPathContextPtr xpath_ctx = xmlXPathNewContext(node->doc);
+		if(xpath_ctx != NULL) {
+			xpath_ctx->node = node;
+			xmlXPathObjectPtr xpath_obj = xmlXPathEvalExpression(BAD_CAST("(.//. | .//@* | .//namespace::*)"), xpath_ctx);
+			xpath_ctx->node = NULL;
+			
+			if (xpath_obj != NULL && xpath_obj->type == XPATH_NODESET) {
+				int c14n_ret = xmlC14NDocSaveTo(node->doc, xpath_obj->nodesetval, XML_C14N_EXCLUSIVE_1_0, NULL, 0, buffer);
+				if (c14n_ret >= 0) {
+					#ifdef LIBXML2_NEW_BUFFER
+						int buffer_size = xmlOutputBufferGetSize(buffer);
+					#else
+						int buffer_size = buffer->buffer->use;
+					#endif
+					if (buffer_size > 0) {
+						#ifdef LIBXML2_NEW_BUFFER
+							char *buffer_content = (char *)xmlOutputBufferGetContent(buffer);
+						#else
+							char *buffer_content = (char *)buffer->buffer->content;
+						#endif
+						result = zend_string_init(buffer_content, strlen(buffer_content), 0);
+					}
+				}
+			}
+
+			xmlXPathFreeObject(xpath_obj);
+			xmlXPathFreeContext(xpath_ctx);
+		}
+		xmlOutputBufferClose(buffer);
+	}
+
+	return result;
+}
+/* }}} */
+
+static int add_reference_to_signed_info(zval *this_ptr, xmlNodePtr signed_info, xmlNodePtr source, zend_string *source_uri, xmlNsPtr ns_ds) /* {{{ */
+{
+	xmlNodePtr reference = xmlNewChild(signed_info, ns_ds, BAD_CAST("Reference"), NULL);
+	xmlSetProp(reference, BAD_CAST("URI"), BAD_CAST(ZSTR_VAL(source_uri)));
+	xmlNodePtr reference_trsfrms = xmlNewChild(reference, ns_ds, BAD_CAST("Transforms"), NULL);
+	xmlNodePtr reference_trsfrm = xmlNewChild(reference_trsfrms, ns_ds, BAD_CAST("Transform"), NULL);
+	xmlSetProp(reference_trsfrm, BAD_CAST("Algorithm"), BAD_CAST(SOAP_WSS_EXC_C14N));
+	xmlNodePtr reference_dm = xmlNewChild(reference, ns_ds, BAD_CAST("DigestMethod"), NULL);
+	
+	const php_hash_ops *ops = NULL;
+	zend_string *algo = NULL;
+	zend_long digest_method = Z_LVAL_P(Z_CLIENT_WSS_DIGEST_METHOD_P(this_ptr));
+	switch (digest_method) {
+		case SOAP_WSS_DIGEST_METHOD_SHA1:			
+			algo = zend_string_init("sha1", strlen("sha1"), 0);
+			ops = php_hash_fetch_ops(algo);
+			xmlSetProp(reference_dm, BAD_CAST("Algorithm"), BAD_CAST(SOAP_WSS_SHA1));
+			break;
```

Comment:
```suggestion
```

SHA1 ought to be dead.  Let’s not add another use of it.

Reply:
It might be needed for legacy servers so I would keep it - there's not such a big harm in it.